### DISCUSSION TITLE
feat: User profile update

### DISCUSSION THREAD

**Discussion by zacjones93** [eggheadio/egghead-next#1212] [General]
I met with @DrShpongle to talk through updating the user profile

There are quite a few features of the old user profile that don't exist on the new site. 

Basic metadata for the user can't be edited other than the primary email for the account. 
We want to add the ability to edit:
- preferred name
- avatar
- connect / disconnect from github (can only connect)

The starting point for the new profile page would be [tailwindui settings page](https://tailwindui.com/components/application-ui/page-examples/settings-screens#component-3e81b8353a7c0ffd711ce35c6b949289). The tabs we see as makes sense is activity, membership, account metadata, and payment.

![user-profile-sketch](https://user-images.githubusercontent.com/6188161/191358168-fdf3f221-d315-4605-9379-5b5ba9484a0c.png)

For logged in users, activity would be a reasonable default for the active tab. This could be something we restrict to PRO members.

The membership page could list out the invoices that we have generated and the individual invoice pages could link back to the membership page.

This structure would set us up for a 'public profile' section that the learner could craft as a future project.

**TO ADD:**
- [ ] "X" button for Continue Learning items (to hide it if user doesn't want to continue with it)
- [ ] Preferred Name field
- [ ] Disconnect from GitHub (?)
- [ ] Upload Avatar functionality

**Comment by ltclmbdev** [comment]
Next steps in this page development can be the following components that require server-side functionality:
![Image 2022-09-30 at 13 33 55](https://user-images.githubusercontent.com/1519448/193352354-8e94f37c-7cea-47f9-b330-22e8ec691608.jpg)

Can we pair on this, @jbranchaud ?

**Reply by jbranchaud** [reply]
I'm taking a look at what we have in egghead-rails to support this.

1. For _Preferred Name_,
  - We'd have to add the `preferred_name` column to the `users` table in the database.
  - We'd also need to update the API to support receiving and setting the `preferred_name` attribute.
  - Then we'd need to determine if there is anywhere in the codebase that we want to replace the use of `first_name` with `preferred_name`.
  - If not, we could just return `preferred_name` in the graphql response as a new attribute and let egghead-next use it wherever it wants.

2. For _Avatar_,
  - This is already a supported attribute of `users` in egghead-rails.
  - We will need to add support for it to the API, because as far as I can tell there isn't a way to update it via the REST API yet.

**Are there other parts of the new user page that need egghead-rails support/updates?**

**Reply by zacjones93** [reply]
I see using `first_name` and returning `preferred_name` in graphql as a good option for this.

**Comment by ltclmbdev** [comment]
Other component I can think about is **Connect to GitHub**: 
![Image 2022-10-04 at 11 05 57](https://user-images.githubusercontent.com/1519448/193893518-5b808fca-d9e3-4352-90b2-11985b2254ae.jpg)
We already have "connect to" functionality, but not "disconnect from".

**Comment by ltclmbdev** [comment]
As one of our users suggests:
"... for the list of "continue learning" and a **button/cross to hide those** i don't want to continue learning and don't want to see in this list".
Sounds like a thing we should apply.
Would it be possible to add some `resource.toggle_continue_learning_url` by analogy with `bookmark.toggle_favorite_url` for the bookmarks list? @jbranchaud
